### PR TITLE
[DRUP 404] Developer Balance: Refresh

### PR DESCRIPTION
**Pull request by shadcn**
Adds caching to prepaid balance and a refresh button to pull latest data.

### CODE REVIEW

File: apigee_m10n.permissions.yml [hunk 1]
```diff
@@ -7,6 +7,9 @@ access monetization packages:
 view mint prepaid reports:
   title: 'View Mint Prepaid Reports'
 
+refresh prepaid balance reports:
```
Comment (cnovak) [inline apigee_m10n.permissions.yml]: I don't think we need a full permission for this, do we?  If you have permission to view balance, you should be able to refresh the balance also.

Comment (Jaesin) [reply apigee_m10n.permissions.yml]: It might be handy. Site admins could disable the feature by not granting anyone the permission to refresh.

Comment (shadcn) [reply apigee_m10n.permissions.yml]: @cnovak @Jaesin can you let me know if we need refresh permissions or not please? I'm doing a refactor. If we are not adding permissions I can remove it while I'm at it. Thanks.

Comment (cnovak) [reply apigee_m10n.permissions.yml]: You can leave it in since @Jaesin said it may be good, thx.  Just wanted to ask to make sure it made sense.

**Review by cnovak — APPROVED** [review]
LGTM

File: src/Form/PrepaidBalanceRefreshForm.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    // Invalidate cache tags.
+    if ($tags = $form_state->get('tags')) {
+      Cache::invalidateTags($tags);
+
+      $this->messenger()->addStatus(static::SUCCESS_MESSAGE);
```
Comment (Jaesin) [inline src/Form/PrepaidBalanceRefreshForm.php]: This should be a translated string.

File: modules/add_credit/src/Job/BalanceAdjustmentJob.php [hunk 3]
```diff
@@ -137,7 +137,7 @@ protected function executeRequest() {
         // total so we have to grab that from the balance controller again.
         $balance_after = $this->getPrepaidBalance($controller, $currency_code);
         $new_balance = new Price((string) ($balance_after->getTopUps()), $currency_code);
-        Cache::invalidateTags([BillingController::CACHE_PREFIX . ':user:' . $this->developer->id()]);
+        Cache::invalidateTags([PrepaidBalanceController::CACHE_PREFIX . ':user:' . $this->developer->id()]);
       }
```
Comment (Jaesin) [inline modules/add_credit/src/Job/BalanceAdjustmentJob.php:141]: Can this use `PrepaidBalanceController::getCacheId($this->developer)`?

I kind of like having the id builder all in one place.

File: src/Controller/PrepaidBalanceController.php [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+      ],
+    ];
+
+    // Cache the render array if enabled.
+    if ($max_age = $this->getCacheMaxAge()) {
+      $build['prepaid_balances']['#cache'] = [
+        'contexts' => ['url.path'],
+        'tags' => $this->getCacheTags($user),
+        'max-age' => $max_age,
+        'keys' => [$this->getCacheId($user, 'prepaid_balances')],
```
Comment (Jaesin) [inline src/Controller/PrepaidBalanceController.php]: Should this be `static::getCacheId()`?

File: src/Controller/PrepaidBalanceController.php [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        return $this->monetization->getSupportedCurrencies();
+      });
+
+      $billing_documents = $this->getDataFromCache($user, 'billing_documents', function () {
+        return $this->monetization->getBillingDocumentsMonths();
+      });
+
+      // Build the form.
+      $build['download_form'] = $this->formBuilder->getForm(PrepaidBalanceReportsDownloadForm::class, $user, $supported_currencies, $billing_documents);
+      $build['download_form']['#cache']['keys'] = [$this->getCacheId($user, 'download_form')];
```
Comment (Jaesin) [inline src/Controller/PrepaidBalanceController.php]: Should this be `static::getCacheId()`?

File: src/Controller/PrepaidBalanceController.php [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+   * @param \Drupal\user\UserInterface $user
+   *   The user entity.
+   *
+   * @return array
+   *   The cache tags.
+   */
+  public static function getCacheTags(UserInterface $user) {
+    return [
+      static::CACHE_PREFIX,
+      static::CACHE_PREFIX . ':user:' . $user->id(),
```
Comment (Jaesin) [inline src/Controller/PrepaidBalanceController.php]: Can `static::CACHE_PREFIX . ':user:' . $user->id()` use `static::getCacheId()` to keep the id building all in one place?

How is `static::CACHE_PREFIX` used? I'm wondering if/why we need more that just `'apigee.monetization.billing:user:' . $user_id`?

Should `apigee.monetization.billing` be changed to `apigee.monetization.prepaid_balance`?

File: src/Form/BillingConfigForm.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+use Drupal\Core\Form\ConfigFormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Config form for billing.
+ *
+ * @package Drupal\apigee_m10n
+ */
+class BillingConfigForm extends ConfigFormBase {
```
Comment (Jaesin) [inline src/Form/BillingConfigForm.php]: Getting nit picky here but `BillingConfigForm` sounds very. similar to the `BillingDetailsForm`.

I think we want to create a monetization settings form with a cache section we can move this one setting to.

**Comment by Jaesin** [comment]
There may be some responses that aren't being queued.

image

File: src/Controller/PrepaidBalanceController.php (hunk 6 above)
Comment (shadcn) [reply src/Controller/PrepaidBalanceController.php]: The `static::CACHE_PREFIX` tag is used to clear all cache when the max_age is updated. See submit handler for the config form.

I've renamed billing to prepaid balance (I was planning to do this in another PR)

File: src/Form/BillingConfigForm.php (hunk 7 above)
Comment (shadcn) [reply src/Form/BillingConfigForm.php]: I renamed the form to `PrepaidBalanceConfigForm`.

**Comment by shadcn** [comment]
> There may be some responses that aren't being queued.
> 
> image

What's the error? Are you testing with integration on?

**Pull request merged**